File: money.py
"""
ID: totheso1
LANG: PYTHON3
TASK: money
"""

from typing import Any
from itertools import chain

fin = open("money.in", "r")
fout = open("money.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
COIN_COUNT, TARGET = map(int, fin.readline().split())
count_collection = [1] + [0] * TARGET
coin_collection = tuple(chain.from_iterable(map(int, line.split()) for line in fin))
for coin in coin_collection:
    for i in range(TARGET + 1):
        if coin <= i:
            count_collection[i] += count_collection[i - coin]
fprint(count_collection[TARGET])
# Code end

File: money.cpp
/*
ID: totheso1
LANG: C++14
TASK: money
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <vector>

using namespace std;

static ifstream fin("money.in");
static ofstream fout("money.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static const int COIN_COUNT = fin_get<int>();
static const int TARGET = fin_get<int>();

int main() {
  vector<unsigned long long> count_collection(TARGET + 1);
  count_collection[0] = 1;
  int coin;
  while (fin >> coin) {
    for (int i = 0; i <= TARGET; ++i) {
      if (coin <= i) {
        count_collection[i] += count_collection[i - coin];
      }
    }
  }
  fout << count_collection[TARGET] << endl;
}
